File: Stage20/C/Problem14888.c
#include <stdio.h>
#include <limits.h>

int N;
int nums[12];
int op[4];
int max = INT_MIN;
int min = INT_MAX;

void dfs(int idx, int current) {
    if (idx == N) {
        if (current > max) max = current;
        if (current < min) min = current;
        return;
    }

    for (int i = 0; i < 4; i++) {
        if (op[i] > 0) {
            op[i]--;

            if (i == 0) {
                dfs(idx + 1, current + nums[idx]);
            } else if (i == 1) {
                dfs(idx + 1, current - nums[idx]);
            } else if (i == 2) {
                dfs(idx + 1, current * nums[idx]);
            } else {
                int next;
                if (current < 0) {
                    next = - ( (-current) / nums[idx] );
                } else {
                    next = current / nums[idx];
                }
                dfs(idx + 1, next);
            }

            op[i]++;
        }
    }
}

int main(void) {
    scanf("%d", &N);

    for (int i = 0; i < N; i++) scanf("%d", &nums[i]);
    for (int i = 0; i < 4; i++) scanf("%d", &op[i]);

    dfs(1, nums[0]);

    printf("%d\n%d\n", max, min);
    return 0;
}

File: Stage20/Python/Problem14888.py
import sys
input = sys.stdin.readline

N = 0
nums = []
op = [0, 0, 0, 0]
max_val = -10**12
min_val = 10**12

def dfs(idx, current):
    global max_val, min_val, nums, op, N

    if idx == N:
        max_val = max(max_val, current)
        min_val = min(min_val, current)
        return

    for i in range(4):
        if op[i] > 0:
            op[i] -= 1

            if i == 0:
                dfs(idx + 1, current + nums[idx])
            elif i == 1:
                dfs(idx + 1, current - nums[idx])
            elif i == 2:
                dfs(idx + 1, current * nums[idx])
            else:
                if current < 0:
                    dfs(idx + 1, -(-current // nums[idx]))
                else:
                    dfs(idx + 1, current // nums[idx])

            op[i] += 1

def main():
    global N, nums, op

    N = int(input())
    nums = list(map(int, input().split()))
    op = list(map(int, input().split()))

    dfs(1, nums[0])

    print(max_val)
    print(min_val)


if __name__ == "__main__":
    main()